qubesfiletrust/qvm_file_trust.py

```diff
@@ -0,0 +1,546 @@
+            'Unable to set original permissions of {}'.format(path))
+        return True
+
+    # Return to original permissions
+    safe_chmod(path, orig_perms,
+        'Unable to set original permissions of {}'.format(path))
+
+    # We didn't find our attribute
+    return False
+
+def set_visual_attributes(path, attributes_on):
+    """Add visual attributes such as emblems and colors to files/folders"""
+    if attributes_on:
+        try:
+            # Add important nautilius emblem
+            proc = subprocess.Popen(['/usr/bin/gvfs-set-attribute', os.path.realpath(path), '-t', 'stringv',
+                'metadata::emblems', 'important'])
+            subprocess.Popen.wait(proc)
+            os.utime(path, None)
+        except:
+            error('Unable to add emblem \'important\' on {}'.format(path))
+
+    else:
+        try:
+            # Remove important nautilus emblem
+            proc = subprocess.Popen(['/usr/bin/gvfs-set-attribute', os.path.realpath(path), '-t', 'unset',
+                'metadata::emblems'], stdout=subprocess.PIPE)
+            subprocess.Popen.wait(proc)
+            os.utime(path, None)
+        except:
+            error('Unable to remove emblem on {}'.format(path))
+
+def path_is_parent(parent, child):
+    """Check if a child file/path is in a parent folder/path."""
+
+    parent = os.path.abspath(parent)
+    child = os.path.abspath(child)
+
+    return os.path.commonpath([parent]) == os.path.commonpath([parent, child])
+
+def is_untrusted_path(path):
+    """Check to see if the path lies under a path that's considered untrusted
+
+    Files listing untrusted paths lie in /etc/qubes/ and ~/.config/qubes
+    under the name always-open-in-dispvm.list
+    """
+
+    untrusted_paths = retrieve_untrusted_folders()
+
+    # Checks if the path is a child of an untrusted path (might be Linux only)
+    for untrusted_path in untrusted_paths:
+        if path_is_parent(untrusted_path, path):
+            return True
+
+    # Check if untrusted phrase (/etc/qubes/always-open-in-dispvm.phrase) is
+    # present in file path
+    global UNTRUSTED_PHRASE
+    return UNTRUSTED_PHRASE.upper() in path.upper()
+
+def check_file(path, multiple_paths):
+    """Check if the given file is trusted"""
+
+    # Save the original permissions of the file.
+    orig_perms = os.stat(path).st_mode
+
+    # See if the file is readable
+    try:
+        with open(path):
+            pass
+
+    except IOError:
+        # Try to unlock file to get read access
+        safe_chmod(path, 0o644,
+            'Could not unlock {} for reading'.format(path))
+
+    # File is readable, attempt to check trusted status
+    global UNTRUSTED_PATH_FOUND
+    global ALL_PATHS_ARE_UNTRUSTED
+    if is_untrusted_xattr(path, orig_perms):
+        # Print out which paths are untrusted if we're checking multiple paths
+        if multiple_paths:
+            qprint('Untrusted: {}'.format(path), False)
+            UNTRUSTED_PATH_FOUND = True
```

Comment:
Better avoid such side-effects and use return value instead (if needed, it can be a tuple or even more complex structure). If nothing else, such side effects make testing harder (you need to reset it between tests).

Reply:
Could you clarify what you mean by 'side-effects'?

Reply:
I believe it is meant: it's preferable for `check_*()` to return whether the thing being checked satisfies the condition or not, e.g. true/false, or perhaps list or tuple of matching things to be checked (like a filter function) rather than printing (side effect of writing stuff to a file descriptor).